File: src/cpp/src/selfplay/selfplay.cpp
#include "c4zero/selfplay/selfplay.hpp"

#include <algorithm>
#include <cmath>
#include <random>
#include <stdexcept>

namespace c4zero::selfplay {
namespace {

struct PendingSample {
  core::Position position;
  std::array<float, core::kNumActions> policy{};
  std::array<std::uint32_t, core::kNumActions> visits{};
  core::Action action = -1;
};

}  // namespace

GeneratedGame generate_game(
    search::Evaluator& evaluator,
    const SelfPlayConfig& config,
    std::uint64_t game_id) {
  search::PuctConfig mcts_config = config.mcts;
  mcts_config.seed = config.seed ^ (0x9E3779B97F4A7C15ULL + (game_id << 6) + (game_id >> 2));
  search::PuctMcts mcts(mcts_config);
  core::Position position = core::Position::empty();
  search::SearchTree tree(position);
  std::vector<PendingSample> pending;
  GeneratedGame game;

  while (!position.is_terminal()) {
    const double temperature = position.ply < config.temperature_sampling_plies ? 1.0 : 0.0;
    search::SearchResult result = mcts.search(tree, evaluator, config.add_root_noise, temperature);
    if (result.selected_action < 0 || !position.is_legal(result.selected_action)) {
      throw std::runtime_error("self-play selected illegal action");
    }
    std::uint32_t visits = 0;
    for (auto count : result.visit_counts) {
      visits += count;
    }
    if (visits != static_cast<std::uint32_t>(config.mcts.simulations_per_move)) {
      throw std::runtime_error("self-play MCTS visit counts do not match simulations");
    }
    pending.push_back(PendingSample{position, result.policy, result.visit_counts, result.selected_action});
    game.completed_simulations += result.completed_simulations;
    game.leaf_evaluations += result.leaf_evaluations;
    game.terminal_evaluations += result.terminal_evaluations;
    game.max_depth = std::max(game.max_depth, result.max_depth);
    game.search_time_ms += result.search_time_ms;
    position = position.play(result.selected_action);
    if (!tree.advance(result.selected_action)) {
      tree = search::SearchTree(position);
    }
  }

  const auto terminal = position.terminal_value();
  if (!terminal.has_value()) {
    throw std::runtime_error("self-play ended without terminal value");
  }

  game.terminal_value = *terminal;
  game.plies = position.ply;
  game.samples.reserve(pending.size());
  for (const PendingSample& sample : pending) {
    const int flips = static_cast<int>(position.ply) - static_cast<int>(sample.position.ply);
    const float value = *terminal * ((flips % 2 == 0) ? 1.0f : -1.0f);
    game.samples.push_back(data::SelfPlaySample::from_position(
        sample.position,
        sample.policy,
        sample.visits,
        value,
        sample.action,
        game_id));
  }
  return game;
}

}  // namespace c4zero::selfplay

File: archive/python_c4az_2026_05_14/src/c4az/selfplay.py
from __future__ import annotations

import random
from dataclasses import dataclass

import numpy as np

from c4az.data import SelfPlaySample
from c4az.game import Position, initial_position
from c4az.mcts import Evaluator, MCTSConfig, PUCTMCTS, choose_action, visit_counts_to_policy


@dataclass(frozen=True, slots=True)
class SelfPlayConfig:
    games: int = 1
    simulations_per_move: int = 128
    c_puct: float = 1.5
    root_dirichlet_alpha: float = 0.3
    root_exploration_fraction: float = 0.25
    temperature_cutoff_ply: int = 12
    seed: int | None = None


@dataclass(frozen=True, slots=True)
class SelfPlayGame:
    samples: list[SelfPlaySample]
    terminal_value: float
    plies: int


def generate_self_play_games(evaluator: Evaluator, config: SelfPlayConfig) -> list[SelfPlayGame]:
    rng = random.Random(config.seed)
    games = []
    for game_id in range(config.games):
        game_seed = None if config.seed is None else config.seed + game_id
        games.append(_play_one_game(evaluator, config, game_id=game_id, seed=game_seed, rng=rng))
    return games


def _play_one_game(
    evaluator: Evaluator,
    config: SelfPlayConfig,
    *,
    game_id: int,
    seed: int | None,
    rng: random.Random,
) -> SelfPlayGame:
    position = initial_position()
    tree = None
    pending: list[tuple[Position, np.ndarray, np.ndarray, int]] = []
    mcts = PUCTMCTS(
        evaluator,
        MCTSConfig(
            simulations_per_move=config.simulations_per_move,
            c_puct=config.c_puct,
            root_dirichlet_alpha=config.root_dirichlet_alpha,
            root_exploration_fraction=config.root_exploration_fraction,
            seed=seed,
        ),
    )
    while not position.is_terminal:
        result = mcts.search(position, root=tree, add_root_noise=True, temperature=1.0)
        temperature = 1.0 if position.ply < config.temperature_cutoff_ply else 0.0
        policy = visit_counts_to_policy(result.visit_counts, temperature, position.legal_mask())
        action = choose_action(policy, rng)
        pending.append((position, policy, result.visit_counts, action))
        tree = mcts.advance_tree(result.root, action)
        position = position.play(action)

    assert position.terminal_value is not None
    samples = []
    final_ply = position.ply
    for sample_position, policy, visits, action in pending:
        flips = final_ply - sample_position.ply
        value = float(position.terminal_value) * ((-1.0) ** flips)
        samples.append(
            SelfPlaySample.from_position(
                sample_position,
                policy=policy,
                value=value,
                visit_counts=visits,
                action=action,
                game_id=game_id,
            )
        )
    return SelfPlayGame(samples=samples, terminal_value=float(position.terminal_value), plies=position.ply)
